gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java

```diff
@@ -0,0 +1,112 @@
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalExtensionOnly;
+import org.threeten.bp.Duration;
+
+/**
+ * Implementations of this class trace the logical flow of a google cloud client.
+ *
+ * <p>A single instance of a tracer represents a logical operation that can be annotated throughout
+ * its lifecycle. Constructing an instance of a subclass will implicitly signal the start of a new
+ * operation.
+ */
+@BetaApi("Surface for tracing is not yet stable")
+@InternalExtensionOnly
+public interface ApiTracer {
+  /**
+   * Asks the underlying implementation to install itself as a thread local. This allows for interop
+   * between clients using gax and external resources to share the same implementation of the
+   * tracing. For example OpenCensus will install a thread local that can read by the GRPC.
+   */
+  Scope inScope();
+
+  /**
+   * Signals that the overall operation has finished successfully. The tracer is now considered
+   * closed and should no longer be used.
+   */
+  void operationSucceeded();
+
+  /**
+   * Signals that the overall operation has failed and no further attempts will be made. The tracer
+   * is now considered closed and should no longer be used.
+   */
+  void operationFailed(Throwable error);
+
+  /** Annotates the operation with selected connection id from the {@code ChannelPool}. */
+  void connectionSelected(int id);
+
+  /**
+   * Adds an annotation that an attempt is about to start. In general this should occur at the very
+   * start of the operation. The attemptNumber is zero based. So the initial attempt will be 0.
+   */
+  void attemptStarted(int attemptNumber);
+
+  /** Adds an annotation that the attempt succeeded. */
+  void attemptSucceeded();
+
+  /**
+   * Adds an annotation that the attempt failed, but another attempt will be made after the delay.
+   */
+  void attemptFailed(Throwable error, Duration delay);
+
+  /**
+   * Adds an annotation that the attempt failed and that no further attempts will be made because
+   * retry limits have been reached.
+   */
+  void attemptFailedRetriesExhausted();
+
+  /**
+   * Adds an annotation that the attempt failed and that no further attempts will be made because
+   * the last error was not retryable.
+   */
+  void attemptPermanentFailure(Throwable error);
+
+  /** Adds an annotation that a streaming response has been received. */
+  void responseReceived();
+
+  /** Adds an annotation that a streaming request has been sent. */
+  void requestSent();
+
+  /** Adds an annotation that a batch of writes has been flushed. */
+  void batchRequestSent(long elementCount, long requestSize);
+
+  /**
+   * A context class to be used with {@link #inScope()} and a try-with-resources block. Closing a
+   * {@link Scope} removes any context that the underlying implementation might've set in {@link
+   * #inScope()}.
+   */
+  interface Scope extends AutoCloseable {
```

Comment:
I don't remember if we've already discussed this, but why do we want to add an abstraction on top of opencensus here? Why not just expose it as is? Opencensus itself already has a lot of power and flexibility. Specifically:
- opencensus is already good in `noop` world (i.e. just calling it and not using the reports is effectively a resource efficient noop tracer already).
- opencensus already has pluggable modules (so we can specify which "reporting system to use" (or whatever that is called properly)) and is flexible to the extent that maybe we do not even need to be able to plug a different reporting mechanism (not opencensus). Exposing opencensus directly can be an option, and this allows us to not have another level of abstraction (like this interface, for example)

Reply:
There are couple of reasons for this approach:
1. I didn't want to sprinkle arbitrary annotation strings throughout the code base. I think it's a lot easier to reason about if all of the annotations are centralized in a single package.
2. I would like to include a fair amount of attributes with the annotations like response counts, these numbers are only needed for tracing and I don't want to pollute business logic at call sites with the accounting. Also some of the attributes cut across layers: like an operation span having an attempt count
3. Finally, an unintentional benefit is that hiding the surface of opencensus gives us a way out in case we get into dependency hell like guava